### PR TITLE
feat: add coorganizer check for creating group

### PR DESCRIPTION
**Pull request by Tlazypanda**
Ref #7597 

#### Short description of what this resolves:
add coorganizer check for creating group only allowing min coorganizer check for the events to be added under a group.

#### Checklist

- [ ] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [ ] My branch is up-to-date with the Upstream `development` branch.
- [ ] The unit tests pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

- [ ] All the functions created/modified in this PR contain relevant docstrings.

### CODE REVIEW

**Comment by iamareebjamal** [comment]
You can add events to a group using patch request to a group as well

**Comment by Tlazypanda** [comment]
@iamareebjamal is this okay now? :sweat_smile:

File: app/api/groups.py [hunk 1]
```diff
@@ -31,6 +32,18 @@ def before_post(self, args, kwargs, data):
...
             raise ForbiddenError({'source': ''}, 'Access Forbidden')
 
+    def before_create_object(self, data, view_kwargs):
+        """
+        before create object method for GroupListPost Class
+        :param data:
+        :param view_kwargs:
+        :return:
+        """
+        if data.get('event', None):
```
Comment (iamareebjamal) [inline app/api/groups.py:42]: Also, shouldn't this be events instead? A group can have multiple events

**Comment by iamareebjamal** [comment]
Please add tests for both cases as well

While creating group
While updating group

**Comment by Tlazypanda** [comment]
@iamareebjamal updated mr :v: can you please review? :sweat_smile: Sorry for the delay got caught up in the package and autoflush issue.

File: app/api/groups.py [hunk 2]
```diff
@@ -115,10 +118,14 @@ def before_update_object(self, group, data, view_kwargs):
...
         """
-        if group and data.get('event'):
-            event = Event.query.filter_by(id=data['event']).first()
-            if event and not has_access('is_coorganizer', event_id=event.id):
-                raise ForbiddenError({'source': ''}, "Event co-organizer access required")
+
+        if 'events' in data:
+            for event in data['events']:
+                event = Event.query.filter_by(id=int(event)).first()
+                if event and not has_access('is_coorganizer', event_id=event.id):
```
Comment (iamareebjamal) [inline app/api/groups.py:125]:
```suggestion
                if not has_access('is_coorganizer', event_id=event):
```

File: app/api/groups.py [hunk 3]
```diff
@@ -115,10 +118,14 @@ def before_update_object(self, group, data, view_kwargs):
...
+                    raise ForbiddenError(
+                        {'source': ''}, "Event co-organizer access required"
+                    )
```
Comment (iamareebjamal) [inline app/api/groups.py:128]:
```suggestion
        for event in data.get('events', []):
            if not has_access('is_coorganizer', event_id=event):
                raise ForbiddenError(
                    {'source': ''}, "Event co-organizer access required"
                )
```

File: app/api/groups.py [hunk 4]
```diff
@@ -39,10 +39,13 @@ def before_create_object(self, data, view_kwargs):
...
-            event = Event.query.filter_by(id=data['event']).first()
-            if event and not has_access('is_coorganizer', event_id=event.id):
-                raise ForbiddenError({'source': ''}, "Event co-organizer access required")
+        if data.get('events', None):
+            for event in data['events']:
+                event = Event.query.filter_by(id=int(event)).first()
+                if event and not has_access('is_coorganizer', event_id=event.id):
+                    raise ForbiddenError(
+                        {'source': ''}, "Event co-organizer access required"
+                    )
```
Comment (iamareebjamal) [inline app/api/groups.py:48]:
```suggestion
        for event in data.get('events', []):
            if not has_access('is_coorganizer', event_id=event):
                raise ForbiddenError(
                    {'source': ''}, "Event co-organizer access required"
                )
```

File: tests/all/integration/api/group/test_group_misc.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    role, _ = get_or_create(Role, name='coorganizer', title_name='Co-organizer')
+    UsersEventsRoles(user=new_user, event=event, role=role)
+
+    db.session.commit()
+    data = json.dumps(
+        {
+            "data": {
+                "type": "group",
+                "relationships": {
+                    "user": {"data": {"id": str(new_user.id), "type": "user"}},
```
Comment (iamareebjamal) [inline tests/all/integration/api/group/test_group_misc.py:25]: User should be dumps only. I should not be able to make you the creator of the group

File: tests/all/integration/api/group/test_group_misc.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+                },
+                "attributes": {"name": "eventgp2"},
+            }
+        }
+    )
+
+    response = client.post(
+        f'/v1/groups',
+        content_type='application/vnd.api+json',
+        headers={'Authorization': 'JWT ' + create_access_token(new_user.id, fresh=True)},
```
Comment (iamareebjamal) [inline tests/all/integration/api/group/test_group_misc.py:36]: No need of creating a user and jwt, just add user and jwt as paramters in the test function, they'll be injected automatically

See other tests for reference

Comment (Tlazypanda) [reply tests/all/integration/api/group/test_group_misc.py:36]: @iamareebjamal sir I had done that initially but by default user is_verified is false so I also tried setting user.is_verified but that was throwing error so I used new user is there any way to bypass this?

**Comment by Tlazypanda** [comment]
@iamareebjamal can you help me with the autoflush issue with the test it comes and goes :sweat_smile: Test is failing because of it if we can retrigger the tests they might pass

**Comment by iamareebjamal** [comment]
The issue is that user_id is null when group is being created. Make user_id dumps_only and set user_id as current user's ID when group is being created

File: app/api/groups.py [hunk 7]
```diff
@@ -31,6 +32,18 @@ def before_post(self, args, kwargs, data):
         if not current_user.is_verified:
             raise ForbiddenError({'source': ''}, 'Access Forbidden')
```
Comment (iamareebjamal) [inline app/api/groups.py:33]: As we are using before_create_object now, move all of it there and remove before_post and delete `data['user'] = current_user.id`

Comment (Tlazypanda) [reply app/api/groups.py:33]: Yes :v:

**Comment by Tlazypanda** [comment]
@iamareebjamal is this okay now? :sweat_smile:

**Pull request merged**